Set up and initialize database

nphilipp opened the pull request:
Fixes: #95 

This comes with setting up the `duffy` CLI to use sub-commands for different things: running the app, setting up the database, applying DB migrations, …

Additionally some generic changes:
- Rename virtualenv directory in GitHub actions
- Make tox runs less redundant
- Add duffy_config fixture and apply globally
- Make isort and black coexist well
- Fix import order

---------

How to test:

* Activate your virtual environment
* Run the test suite 😉
  ```
  tox
  ```

Manual tests:

* Copy the example configuration file…
  ```
  cp etc/duffy-example-config.yaml etc/duffy-config.yaml
  ```
* … and edit `database.sqlalchemy.*_url` to point to a real file, e.g. `devdata.db`:
  ```
  database:
    sqlalchemy:
      sync_url: "sqlite:///devdata.db"
      async_url: "sqlite+aiosqlite:///devdata.db"
  ```
* Set up the database, i.e. create tables from the DB model
  ```
  duffy -c etc/duffy-config.yaml setup-db
  ```
  The configured database file `devdata.db` should contain tables corresponding to what's defined in `duffy.database.model`, you can verify that in `sqlite3` with the command `.tables` or using the `sqliteman` GUI. 
* Run the app to serve the API:
  ```
  duffy --config etc/duffy-config.yaml serve --reload
  ```
  It should serve requests on http://127.0.0.1:8080

nphilipp committed:
Move cli module out of duffy.app

We'll have subcommands soon, and most of them aren't about running the
web app.

Signed-off-by: Nils Philippsen <nils@redhat.com>
Signed-off-by: siddharthvipul <siddharthvipul1@gmail.com>

nphilipp committed:
Fix import order

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Add serve subcommand

This runs the web app and allows us to add other subcommands, e.g. to
migrate the database schema. Also, test return code of the command.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Straighten out sync vs. async DB access

Previously, I assumed that we could base everything regarding database
engines and sessions off of SQLAlchemy's AsyncEngine. However, I didn't
manage to bind its .sync_engine to a synchronous Session... For the time
being at least, we'll configure and initialize the sync and async
engines and sessions separately.

Additionally, require Python >= 3.7 because we need asyncio.current_task
which isn't available in 3.6 and on Python 3.7, disable the pytest
coverage plugin because some tests have to be skipped there.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Make isort and black coexist well

Previously, both tools disagreed how to format multi-line imports.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Add a command to set up the database schema

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Add duffy_config fixture and apply globally

This ensures that every test starts with a clean configuration. Like the
duffy_config_files fixture on which it depends, custom configuration can
be injected using the marker of the same name,
@pytest.mark.duffy_config().

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Make tox runs less redundant

Don't run black, flake8, isort test during normal pytest runs, add their
own environments for this.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Rename virtualenv directory in GitHub actions

This caused flake8 to check all our dependencies before, so made its
report useless.

Signed-off-by: Nils Philippsen <nils@redhat.com>

nphilipp committed:
Initialize DB model when starting the app

Also, rewrap errors due to missing configuration keys and present them
more nicely to users.

Signed-off-by: Nils Philippsen <nils@redhat.com>

gridhead reviewed (changes requested):
Please provide instructions on checking/reviewing this PR.

Also, this is what I see when I execute the following command.
```
(venv) [t0xic0der@fedorable duffy]$ duffy setup-db
```

```
Traceback (most recent call last):
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/bin/duffy", line 5, in <module>
    cli()
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/duffy/cli.py", line 64, in setup_db
    database.setup.setup_db_schema()
AttributeError: module 'duffy.database' has no attribute 'setup'
```

Python 3.10.0
Fedora 35

gridhead commented on duffy/cli.py:
```diff
@@ -3,14 +3,18 @@
 
-from ..configuration import config, read_configuration
-from ..version import __version__
+from . import database
```
```suggestion
from duffy.database.setup import setup_db_schema
from duffy.database import init_model
```

nphilipp replied:
I want to stick to relative imports within the package (`migrations/env.py` is an exception by way of how it's invoked by alembic, maybe this needs to be commented), and prefixing it with `database` is intentional, it saves me from typing a comment down below (about what model this call is about). I need to think about how best to do it, though.

gridhead replied:
The changes that I listed in this thread of subcomments are "not" suggestions but rather, the things that I did to make the code work and continue reviewing the pull request. :)

nphilipp replied:
👍

gridhead commented on duffy/cli.py:
```diff
@@ -45,26 +44,64 @@ def init_config(ctx, param, filename):
+@cli.command()
+def setup_db():
+    """Create tables from the database model."""
+    database.setup.setup_db_schema()
```
```suggestion
    setup_db_schema()
```

nphilipp replied:
See above.

gridhead commented on duffy/cli.py:
```diff
@@ -45,26 +44,64 @@ def init_config(ctx, param, filename):
         uvicorn_log_config["loggers"]["duffy"]["level"] = numeric_loglevel
 
-    logging.basicConfig(level=numeric_loglevel)
+    database.init_model()
```
```suggestion
    init_model()
```

nphilipp replied:
Ditto.

gridhead reviewed (commented):
I played around with the code and made some changes.

I ended up getting the following now.

```
(venv) [t0xic0der@fedorable duffy]$ duffy setup-db
```

```
Traceback (most recent call last):
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/bin/duffy", line 5, in <module>
    cli()
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/duffy/cli.py", line 65, in setup_db
    setup_db_schema()
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/duffy/database/setup.py", line 11, in setup_db_schema
    engine = get_sync_engine()
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/duffy/database/__init__.py", line 68, in get_sync_engine
    sync_config = deepcopy(config["database"]["sqlalchemy"])
KeyError: 'database'
```

(Which changes you ask? The one that I added as review subcomments under this)

gridhead commented:
My bad.

I think my braincells have disappeared.

```
(venv) [t0xic0der@fedorable duffy]$ duffy -c etc/duffy-example-config.yaml setup-db
```
```
Creating database schema
```

gridhead reviewed (changes requested):
With reference to https://github.com/CentOS/duffy/pull/97#issuecomment-972919194, one suggestion that I would like to make here. Please make the following work.

```
(venv) [t0xic0der@fedorable duffy]$ duffy setup-db -c etc/duffy-example-config.yaml 
```

```
Usage: duffy setup-db [OPTIONS]
Try 'duffy setup-db --help' for help.

Error: No such option: -c
```

gridhead reviewed (changes requested):
Let us detect the absence of the config file a bit more elegantly.

Like

```
(venv) [t0xic0der@fedorable duffy]$ duffy run-app
```
```
 * Starting Duffy...
 * Host address : 127.0.0.1
 * Port number  : 8080
 * Log level    : warning
Traceback (most recent call last):
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/bin/duffy", line 5, in <module>
    cli()
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/click/decorators.py", line 26, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/duffy/cli.py", line 105, in run_app
    init_model()
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/duffy/database/__init__.py", line 63, in init_model
    init_sync_model(sync_engine)
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/duffy/database/__init__.py", line 50, in init_sync_model
    sync_engine = get_sync_engine()
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/duffy/database/__init__.py", line 68, in get_sync_engine
    sync_config = deepcopy(config["database"]["sqlalchemy"])
KeyError: 'database'
```

to be changed to 

```
(venv) [t0xic0der@fedorable duffy]$ duffy run-app
```
```
 * Starting Duffy...
 * Host address : 127.0.0.1
 * Port number  : 8080
 * Log level    : warning
 * Config file not provided. Exiting...
```

The same goes for the `setup-db` command.

gridhead commented:
With reference to https://github.com/CentOS/duffy/pull/97#issuecomment-972919194 and just like https://github.com/CentOS/duffy/pull/97#pullrequestreview-810005210, another suggestion that I would like to make here. Please make the following work.

```
(venv) [t0xic0der@fedorable duffy]$ duffy run-app -c etc/duffy-example-config.yaml 
```

```
Usage: duffy setup-db [OPTIONS]
Try 'duffy setup-db --help' for help.

Error: No such option: -c
```

The following works just fine.

```
(venv) [t0xic0der@fedorable duffy]$ duffy -c etc/duffy-example-config.yaml run-app
```

```
 * Starting Duffy...
 * Host address : 127.0.0.1
 * Port number  : 8080
 * Log level    : warning
```

nphilipp commented:
> With reference to [#97 (comment)](https://github.com/CentOS/duffy/pull/97#issuecomment-972919194) and just like [#97 (review)](https://github.com/CentOS/duffy/pull/97#pullrequestreview-810005210), another suggestion that I would like to make here. Please make the following work.

Hmm, from [the documentation of Click](https://click.palletsprojects.com/en/8.0.x/commands/#passing-parameters):

> Click strictly separates parameters between commands and subcommands. What this means is that options and arguments for a specific command have to be specified *after* the command name itself, but *before* any other command names.

=> Supporting global options at arbitrary places in the command line would mean having to repeat them, i.e decorate every subcommand function accordingly. It would also mean that you could specify one config file *before* and one *after* the subcommand and it's unclear what would happen in that case. If we want to allow this, let's use an option parser that supports it natively, e.g. argparse.

gridhead commented:
To https://github.com/CentOS/duffy/pull/97#issuecomment-973394552, fair enough - let us document the same as it is something which is very Click-specific and might not be implied by the admins and/or anyone else who would be interacting with the CLI.

nphilipp committed:
Add custom exceptions for Duffy

This is to enable better error messages (rather than exceptions being
spat onto the console if e.g. some configuration key is missing).

Signed-off-by: Nils Philippsen <nils@redhat.com>

gridhead reviewed (changes requested):
Let's have an elegant exiting out.

```
(venv) [t0xic0der@fedorable duffy]$ duffy --config etc/duffy-config.yaml serve --reload
```
```
 * Starting Duffy...
 * Host address : 127.0.0.1
 * Port number  : 8080
 * Log level    : warning
 * Serving API docs on http://127.0.0.1:8080/docs
^CERROR:    Traceback (most recent call last):
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/starlette/routing.py", line 624, in lifespan
    await receive()
  File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/uvicorn/lifespan/on.py", line 135, in receive
    return await self.receive_queue.get()
  File "/usr/lib64/python3.10/asyncio/queues.py", line 159, in get
    await getter
asyncio.exceptions.CancelledError
```

I can totally that it can be different pain with async programming.

nphilipp commented:
> Let's have an elegant exiting out.
> 
> ```
> (venv) [t0xic0der@fedorable duffy]$ duffy --config etc/duffy-config.yaml serve --reload
> ```
> 
> ```
>  * Starting Duffy...
>  * Host address : 127.0.0.1
>  * Port number  : 8080
>  * Log level    : warning
>  * Serving API docs on http://127.0.0.1:8080/docs
> ^CERROR:    Traceback (most recent call last):
>   File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/starlette/routing.py", line 624, in lifespan
>     await receive()
>   File "/home/t0xic0der/Projects/forks/nphilipp/duffy/venv/lib/python3.10/site-packages/uvicorn/lifespan/on.py", line 135, in receive
>     return await self.receive_queue.get()
>   File "/usr/lib64/python3.10/asyncio/queues.py", line 159, in get
>     await getter
> asyncio.exceptions.CancelledError
> ```
> 
> I can totally that it can be different pain with async programming.

No idea, but it's out of scope for this PR :wink:. I imagine it's something easy like having to catch `KeyboardInterruptError`.

gridhead reviewed (approved):
To https://github.com/CentOS/duffy/pull/97#issuecomment-974121011, fair enough.

LGTM. Merging.
